### PR TITLE
Fix getting partition metadata of a nonexistent topic returns 0

### LINKED ISSUE #8813: Get partition metadata for a non-existed topic returns 0
**Describe the bug**
When getting the partition metadata for a non-existed topic, it returns 0 instead of throwing an exception.

**To Reproduce**
1. Run a pulsar standalone
2. Get a topic's partition number

```java
        try (PulsarAdmin admin = PulsarAdmin.builder().serviceHttpUrl("http://localhost:8080").build()) {
            System.out.println(admin.topics().getPartitionedTopicMetadata("xxxxx").partitions);
        } catch (/* ... */) {
```

3. The output is 0 and no exception was thrown.

**Expected behavior**
It should throw an exception because for a non-partitioned topic, its partitions should be 0, so we cannot distinguish a non-existed topic from a non-partitioned topic.

### PR DESCRIPTION
**Pull request by BewareMyPower**
Fixes #8813 

### Motivation

Actually #8818 didn't fix the issue. The reason is when `PulsarAdmin` tries to get partition metadata, the `checkAllowAutoCreation` query param is false, so the topic existence check will never be performed.

### Modifications

- Check if the topic exists when the partition's metadata is 0 and `checkAllowAutoCreation` is false. See comments for detail explanations.
- Change the current tests to make sure the check works.

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

This change is already covered by existing tests, such as *AdminApiTest#partitionedTopics*.

### CODE REVIEW

**Comment by BewareMyPower** [comment]
Mark it as WIP first because there may be still some failed tests.

**Comment by BewareMyPower** [comment]
Now it's ready for review. It looks like there're still two flaky tests that can always pass in my local env.
- LoadBalancerTest.testBrokerRanking
- MessagePublishBufferThrottleTest.testBlockByPublishRateLimiting

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/AdminApiTest.java [hunk 1]
```diff
@@ -873,6 +872,13 @@ public void persistentTopics(String topicName) throws Exception {
     public void partitionedTopics(String topicName) throws Exception {
         assertEquals(admin.topics().getPartitionedTopicList("prop-xyz/ns1"), Lists.newArrayList());
         final String partitionedTopicName = "persistent://prop-xyz/ns1/" + topicName;
+
+        try {
+            admin.topics().getPartitionedTopicMetadata(partitionedTopicName);
+            fail("getPartitionedTopicMetadata of " + partitionedTopicName + " should not succeed");
+        } catch (NotFoundException ignored) {
```
Comment (eolivelli) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/AdminApiTest.java:879]: nit: rename `ignored` to `expected`

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/AdminApiTest.java [hunk 2]
```diff
@@ -882,8 +888,11 @@ public void partitionedTopics(String topicName) throws Exception {
         List<String> topics = admin.topics().getList("prop-xyz/ns1");
         assertEquals(topics.size(), 4);
 
-        assertEquals(admin.topics().getPartitionedTopicMetadata("persistent://prop-xyz/ns1/ds2").partitions,
-                0);
+        try {
+            admin.topics().getPartitionedTopicMetadata("persistent://prop-xyz/ns1/ds2");
+            fail("getPartitionedTopicMetadata of persistent://prop-xyz/ns1/ds2 should not succeed");
+        } catch (NotFoundException ignored) {
```
Comment (eolivelli) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/AdminApiTest.java:894]: nit: rename `ignored` to `expected`

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/AdminApiTest.java [hunk 3]
```diff
@@ -1003,7 +1012,11 @@ public void partitionedTopics(String topicName) throws Exception {
 
         admin.topics().deletePartitionedTopic(partitionedTopicName);
 
-        assertEquals(admin.topics().getPartitionedTopicMetadata(partitionedTopicName).partitions, 0);
+        try {
+            admin.topics().getPartitionedTopicMetadata(partitionedTopicName);
+            fail("getPartitionedTopicMetadata of " + partitionedTopicName + " should not succeed");
+        } catch (NotFoundException ignored) {
```
Comment (eolivelli) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/AdminApiTest.java:1018]: nit: rename `ignored` to `expected`

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/v1/V1_AdminApiTest.java [hunk 4]
```diff
@@ -792,9 +792,11 @@ public void partitionedTopics(String topicName) throws Exception {
         List<String> topics = admin.topics().getList("prop-xyz/use/ns1");
         assertEquals(topics.size(), 4);
 
-        assertEquals(
-                admin.topics().getPartitionedTopicMetadata("persistent://prop-xyz/use/ns1/ds2").partitions,
-                0);
+        try {
+            admin.topics().getPartitionedTopicMetadata("persistent://prop-xyz/use/ns1/ds2");
+            fail("getPartitionedTopicMetadata of persistent://prop-xyz/use/ns1/ds2 should not succeed");
+        } catch (NotFoundException ignored) {
```
Comment (eolivelli) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/v1/V1_AdminApiTest.java:798]: nit: rename `ignored` to `expected`

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/v1/V1_AdminApiTest.java [hunk 5]
```diff
@@ -908,7 +910,11 @@ public void partitionedTopics(String topicName) throws Exception {
 
         admin.topics().deletePartitionedTopic(partitionedTopicName);
 
-        assertEquals(admin.topics().getPartitionedTopicMetadata(partitionedTopicName).partitions, 0);
+        try {
+            admin.topics().getPartitionedTopicMetadata(partitionedTopicName);
+            fail("getPartitionedTopicMetadata of " + partitionedTopicName + " should not succeed");
+        } catch (NotFoundException ignored) {
```
Comment (eolivelli) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/v1/V1_AdminApiTest.java:916]: nit: rename `ignored` to `expected`

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/BrokerServiceAutoTopicCreationTest.java [hunk 6]
```diff
@@ -154,8 +154,10 @@ public void testGetPartitionedMetadataWithoutCheckAllowAutoCreation() throws Exc
         pulsar.getConfiguration().setDefaultNumPartitions(3);
 
         final String topicString = "persistent://prop/ns-abc/test-topic-3";
-        int partitions = admin.topics().getPartitionedTopicMetadata(topicString).partitions;
-        assertEquals(partitions, 0);
+        try {
+            admin.topics().getPartitionedTopicMetadata(topicString);
+        } catch (PulsarAdminException.NotFoundException ignored) {
```
Comment (eolivelli) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/BrokerServiceAutoTopicCreationTest.java:159]: nit: rename `ignored` to `expected`

**Review by eolivelli — CHANGES_REQUESTED** [review]
awesome work.
can you please also verify the behaviour for non-persistent topics ?
it happens quite often that we make a fix for persistent topics but we let non-persistent topics behave in a inconsistent way

**Comment by BewareMyPower** [comment]
Thanks @eolivelli for your advice. I tried to add tests for non-persistent topics and found some difference between persistent topics and non-persistent topics. In this PR, I'll only fix the `getPartitionedTopicMetadata` implementation of non-persistent topics and mark the difference as comments.

**Comment by eolivelli** [comment]
@BewareMyPower can you please follow up with a new PR for non persistent topics ?
I am fine with merging this patch (but please check my comments)

**Comment by BewareMyPower** [comment]
@eolivelli I'll push a commit to fix the `getPartitionedTopicMetadata` for non-persistent topics in this PR. The other problems are subscription related admin operations. I may not have time to fix them recently, maybe we can open an issue to track it first.

**Comment by BewareMyPower** [comment]
@codelipenghui Please include this fix in 2.8.0 if it's merged

**Review by sijie — APPROVED** [review]

**Comment by BewareMyPower** [comment]
@eolivelli PTAL again

**Review by eolivelli — APPROVED** [review]
great work

**Comment by rdhabalia** [comment]
This PR breaks all previous clients who would like to use HTTP lookup. HTTP lookup at client side expects partitioned metadata with 0 partition and this PR gives `TopicNotFoundException` to client and old client couldn't handle it and won't be able to create a new topic which breaks topic creation semantics of Pulsar.
I have seen multiple such instances in past where we broke the compatibility and it's not acceptable.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 7]
```diff
@@ -543,6 +543,21 @@ protected PartitionedTopicMetadata internalGetPartitionedMetadata(boolean author
...
         PartitionedTopicMetadata metadata = getPartitionedTopicMetadata(topicName,
                 authoritative, checkAllowAutoCreation);
+        if (metadata.partitions == 0 && !checkAllowAutoCreation) {
+            // The topic may be a non-partitioned topic, so check if it exists here.
+            // However, when checkAllowAutoCreation is true, the client will create the topic if it doesn't exist.
+            // In this case, `partitions == 0` means the automatically created topic is a non-partitioned topic so we
+            // shouldn't check if the topic exists.
+            try {
+                if (!pulsar().getNamespaceService().checkTopicExists(topicName).get()) {
+                    throw new RestException(Status.NOT_FOUND,
```
Comment (rdhabalia) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:553]: this breaks HTTP lookup for all the previous clients and also breaks the semantic of topic creation. why are we creating new semantics by breaking the compatibility? what are we trying to achieve here?

**Comment by eolivelli** [comment]
@rdhabalia 
I agree that this is a behaviour change in the API (although we introduced it in a major release, 2.8.0)

I think that this is fixing a bug, if the topic does not exist we should return "not found".

I understand that this is painful for clients who expected "0" instead of "Not Found", we should have advertised this change better or dealt with it in a different way.

Actually at the time of 2.8.0 we didn't have good tools/processes to deal with this kind of breaking changes (generally speaking breaking changes shouldn't be allowed at all!) .

Now the community evolved and before doing this kind of change we have to follow  a more strict process (PIP) and that's should allow us to not fall anymore in this kind of problems for existing users.

**Comment by rdhabalia** [comment]
I am going to revert this change as old clients are just failing. let me know if anyone has any objections. I have experienced such intentional changes in past as well and I would like to check if anyone thinks that it's fine to do destructive changes just because their PRs will be merged then it's not a good practice. So, this time I want to set the message clear for these intentionally breaking changes.

**Comment by eolivelli** [comment]
This change has been released with 2.8 that now is a  very old release (very close to be considered obsolete). I am aware of many new users since 2
8 and for them we will introduce another breaking change.

I think that you could send out PR and start a discussion on dev@

**Comment by rdhabalia** [comment]
before this change, the previous behavior was there for 8 years so, I definitely reject the claim that it's been there for the last 2 years. And again: this change is breaking the lookup for an old clients. it's a P0 and it's not acceptable with any justification.

**Comment by rdhabalia** [comment]
also I don't understand why do we defend breaking changes.? what's the main motivation here? does anyone want me to find out all past breaking change list and we have to live with those changes by patching on top of open source?

**Comment by BewareMyPower** [comment]
This PR was included in 2.8.0 and didn't break the compatibility with 2.7.x because https://github.com/apache/pulsar/pull/4963 introduced a query param `checkAllowAutoCreation` in HTTP client.  Actually #4963 introduced a breaking change (`allowAutoTopicCreationType=partitioned`) so that the query param was added. The breaking change has been reverted in https://github.com/apache/pulsar/pull/5308 (`allowAutoTopicCreationType=non-partitioned`), but the client side changes are not reverted.

The C++ client tests should expose the breaking change because there was no `checkAllowAutoCreation` in C++ client before https://github.com/apache/pulsar/pull/11557. Unfortunately, the CI in master was accidentally broken at that time (see https://github.com/apache/pulsar/pull/11575) so the C++ client tests passed.

I agree that it's a breaking change. But this new behavior is also needed in some cases to differ a non-partitioned topic and a non-existent topic.

I prefer to add a new config to keep the compatibility for old version clients. Reverting this PR would also bring many breaking changes for those external systems that depend on Pulsar 2.8.0 or later. @rdhabalia @eolivelli 

I will send an mail to dev and reference the discussion here soon.

**Comment by BewareMyPower** [comment]
Let's continue the discussion here: https://lists.apache.org/thread/88t1xxf68j092k09srdwyzj1tk4ml5n9

**Pull request merged**